dcm4che-net-audit/src/test/java/org/dcm4che3/net/audit/AuditLoggerTest.java

```diff
@@ -328,17 +353,263 @@ public void reconfigure_UpdatesConnections_WhenCalled() {
+        }
+    }
+
+    @Test
+    public void sendMessage_ThrowsIOException_WhenTwoTLSARRDevicesFail()
+            throws IncompatibleConnectionException, GeneralSecurityException, IOException {
+        AuditMessage msg = new AuditMessage();
+        msg.setEventIdentification(AuditMessages.createEventIdentification(
+                AuditMessages.EventID.ApplicationActivity, AuditMessages.EventActionCode.Execute,
+                Calendar.getInstance(), AuditMessages.EventOutcomeIndicator.Success, null,
+                AuditMessages.EventTypeCode.ApplicationStart));
+
+        final Connection auditConnTLS = createNiceMock(Connection.class);
+        auditConnTLS.setTlsCipherSuites(CIPHER);
+        auditConnTLS.setPort(AUDIT_PORT_TLS);
+        expect(auditConnTLS.getProtocol()).andStubReturn(SYSLOG_TLS);
+        expect(auditConnTLS.getDevice()).andStubReturn(mockHubDevice);
+        expect(auditConnTLS.isInstalled()).andStubReturn(true);
+        expect(auditConnTLS.isCompatible(auditConnTLS)).andStubReturn(true);
+        expect(auditConnTLS.connect(auditConnTLS)).andThrow(new IOException());
+
+        final Connection auditConnNoTLS = createNiceMock(Connection.class);
+        auditConnNoTLS.setPort(AUDIT_PORT_NO_TLS);
+        expect(auditConnNoTLS.getProtocol()).andStubReturn(SYSLOG_TLS);
+        expect(auditConnNoTLS.getDevice()).andStubReturn(mockSyslogDevice);
+        expect(auditConnNoTLS.isInstalled()).andStubReturn(true);
+
+        expect(auditConnTLS.isCompatible(auditConnNoTLS)).andStubReturn(true);
+        expect(auditConnTLS.connect(auditConnNoTLS)).andThrow(new IOException());
+
+        replay(auditConnTLS, auditConnNoTLS);
+
+        final AuditRecordRepository arr1 = new AuditRecordRepository();
+        arr1.addConnection(auditConnTLS);
+        mockHubDevice.addDeviceExtension(arr1);
+        mockHubDevice.addConnection(auditConnTLS);
+        mockHubDevice.setDeviceName(INTERNAL_DEVICE_NAME);
+
+        logger.addAuditRecordRepositoryDevice(mockHubDevice);
+        logger.addConnection(auditConnTLS);
+
+        final AuditRecordRepository arr2 = new AuditRecordRepository();
+        arr2.addConnection(auditConnNoTLS);
+        mockSyslogDevice.addDeviceExtension(arr2);
+        mockSyslogDevice.addConnection(auditConnNoTLS);
+        mockSyslogDevice.setDeviceName(EXTERNAL_DEVICE_NAME);
+
+        logger.addAuditRecordRepositoryDevice(mockSyslogDevice);
+
+        try {
+            logger.write(Calendar.getInstance(), msg);
+            fail("Expected an IOException, but no exception was thrown");
+        } catch (IOException e) {
+            assertThat("Unexpected exception message", e.getMessage(),
+                    is("Unable to send audit message to device name(s) " + INTERNAL_DEVICE_NAME + ", " + EXTERNAL_DEVICE_NAME));
+        }
+    }
+
+    @Test
+    public void sendMessage_ThrowsIOException_WhenOneTLSARRDeviceSucceedsAndOneUDPARRDeviceFails()

```

Comment:
Would it be worth adding a test with the hub connection failing and the external connection succeeding so that both combinations are covered?